generator/c2i/c2i_generator.c2

```diff
@@ -16,53 +16,73 @@
 module c2i_generator;
 
 import ast local;
+import component;
 import console;
 import file_utils;
 import string_buffer;
 import string_list;
+import string_pool;
 
 type Generator struct {
+    string_pool.Pool* astPool;
     string_buffer.Buf* out;
+    const char* output_dir;
     string_list.List imports;  // used to filter out duplicate imports (from different ASTs)
     Module* mod;
+    const char* mod_name;
     bool in_body;
 }
 
-public fn void generate(const char* output_dir,
-                        ast.Module* mod)
+public fn void generate(string_pool.Pool* astPool, const char* output_dir, component.List* comps)
 {
-    Generator gen = {
-        .out = string_buffer.create(256*1024, false, 3),
-        .mod = mod,
-        .in_body = false,
+    Generator gen = {}
+    gen.astPool = astPool;
+    gen.out = string_buffer.create(64*1024, false, 3);
+    gen.output_dir = output_dir;
+
+    component.Component* mainComp = comps.getLast();
+
+    // generate external components
+    for (u32 i=0; i<comps.size(); i++) {
+        component.Component* c = comps.get(i);
+        c.visitModules(Generator.on_module, &gen);
     }
+    // generate main module
+    mainComp.visitModules(Generator.on_module, &gen);
+    gen.out.free();
+}
+
+fn void Generator.on_module(void* arg, Module* m) {
+    if (!m.isUsed() || !m.isExported()) return;
```

Comment:
If you are exporting a module, it probably is a top-level module that is not 'used'